Final size with risk groups using Newton solver

pratikunterwegs opened the pull request:
This PR adds the Newton solver option to `final_size_grps` in R.

1. Implements the Newton solver from https://gitlab.com/epidemics-r/code_snippets/-/blob/feature/newton_solver/include/finalsize.hpp in R, and fixes #55.
2. Adds tests for the Newton solver:
 - Simple cases with synthetic data, fixes #56
 - Checks for failures at higher `r0`, fixes #57
 - Cases with multiple risk groups, fixes #58 
 - Checks for failures with `socialmixr::polymod` data, fixes #59 
3. Adds documentation, fixes #60 
4. Adds `solver` argument to `final_size_grps` to allow users to choose between solvers, fixes #61, fixes #62

pratikunterwegs commented:
Adding a commit to correct the same `r0 = 12` test, which I fixed in https://github.com/epiverse-trace/finalsize/pull/65/commits/30e34864e73e5e78ebb250846603edc3c4be2ccd for the iterative solver only

pratikunterwegs commented:
Updated branch with a local rebase on `main` after merging #63, there were conflicts in the test for the iterative solver with varying `r0`. I resolved these and force pushed the local to origin to avoid duplicate commits.

Bisaloo reviewed (commented):
Glad to see you got it to work :smile:!

I would also add a test to check that the outputs from both solver are identical (or at least, equal with a certain tolerance)

Bisaloo commented on R/finalsize_risk_grps.R:
```diff
@@ -84,18 +86,30 @@ final_size_grps <- function(contact_matrix,
+      tolerance = tolerance
+    )
+  } else {
+    stop("Error: solver must be 'iterative' or 'newton'")
```
It would be better as a `match.arg()` at the beginning of the function.

Bisaloo commented on R/finalsize_risk_grps.R:
```diff
@@ -84,18 +86,30 @@ final_size_grps <- function(contact_matrix,
-
+  epi_final_size <- numeric()
+  # solve for final size using iterative solver
+  if (solver == "iterative") {
+    epi_final_size <- solve_final_size_iterative(
+      contact_matrix = epi_spread_data[["contact_matrix"]],
+      demography_vector = epi_spread_data[["demography_vector"]],
+      p_susceptibility = epi_spread_data[["p_susceptibility"]],
+      susceptibility = epi_spread_data[["susceptibility"]],
+      iterations = iterations,
+      tolerance = tolerance,
+      step_rate = step_rate,
+      adapt_step = adapt_step
+    )
+  } else if (solver == "newton") {
+    epi_final_size <- solve_final_size_newton(
+      contact_matrix = epi_spread_data[["contact_matrix"]],
+      demography_vector = epi_spread_data[["demography_vector"]],
+      susceptibility = epi_spread_data[["susceptibility"]],
+      iterations = iterations,
+      tolerance = tolerance
+    )
```
If / when the arguments to the solver are passed as `control`, this could become:

```
f_solver <- switch(solver,
  iterative = solve_final_size_iterative,
  newton = solve_final_size_newton
)
epi_final_size <- solve_final_size_newton(
  contact_matrix = epi_spread_data[["contact_matrix"]],
  demography_vector = epi_spread_data[["demography_vector"]],
  susceptibility = epi_spread_data[["susceptibility"]],
  control = control
)
```

to remove redundancy in the code and simplify a potential future renaming of the arguments.

pratikunterwegs replied:
Done in https://github.com/epiverse-trace/finalsize/pull/65/commits/891ed778bc79ab9e8d5697b5feee6da02fd720f0

pratikunterwegs commented on R/finalsize_risk_grps.R:
```diff
@@ -84,17 +86,31 @@ final_size_grps <- function(contact_matrix,
+      susceptibility = epi_spread_data[["susceptibility"]],
+      iterations = iterations,
+      tolerance = tolerance
+    )
```
Done in https://github.com/epiverse-trace/finalsize/pull/65/commits/891ed778bc79ab9e8d5697b5feee6da02fd720f0

pratikunterwegs commented on R/newton_solver.R:
```diff
@@ -0,0 +1,82 @@
+  }
+
+  # cache_m is the contact matrix. must not be modified!
+  dx_f <- function(x, cache, cache_m) {
```
Done in https://github.com/epiverse-trace/finalsize/pull/65/commits/b4d840934d5dd7629e53bd7dfca3af3cd360b432

pratikunterwegs commented on R/newton_solver.R:
```diff
@@ -0,0 +1,82 @@
+  epi_final_size <- rep(0.5, n_dim)
+
+  # replicate contact matrix
+  contact_matrix_ <- contact_matrix
```
Done in https://github.com/epiverse-trace/finalsize/pull/65/commits/b4d840934d5dd7629e53bd7dfca3af3cd360b432

pratikunterwegs commented:
Thanks @Bisaloo, I've changed how the `solver` option switches the solver function per your recommendation, and am now passing the solver options as a `control` list.

I removed the extra function arguments remaining from the C++ code - but also the `p_susceptibility` argument from the iterative solver and from the `epi_spread` return. The `epi_spread` function was returning a vector of `1.0`s, which was being used in a multiplication step in `solve_final_size_iterative` - unnecessary. This means both solvers now take the same epi arguments; this is also newly documented.

Bisaloo commented on R/finalsize_risk_grps.R:
```diff
@@ -76,6 +101,31 @@ final_size_grps <- function(contact_matrix,
+  )
+
+  # prepare default list of control params
+  con <- list(
+    iterations = 1000,
+    tolerance = 1e-6,
+    step_rate = 1.9,
+    adapt_step = TRUE
+  )
+  # assign user specified values
+  con[(names(control))] <- control
+  extra_names <- names(control)[!names(control) %in% names(con)]
+  if (length(extra_names)) {
+    warning("Unknown names in control: ", paste(extra_names, collapse = ", "))
+  }
+
+  # prune control list if using newton solver
+  if (solver == "newton") con <- con[c("iterations", "tolerance")]
```
I think it'll be easier to handle the default settings and argument checking directly in the solver functions.

Bisaloo replied:
I just scrolled down a bit more and it looks like this is already done so I believe L112 to 127 can safely be removed.

pratikunterwegs replied:
I'm not sure what you mean, but commenting out L112 - 127 and substituting `control` for `con` at L.147 does pass tests so I'll change that.

Bisaloo commented on R/newton_solver.R:
```diff
@@ -0,0 +1,80 @@
+  # prepare an initial value from where to
+  x_ <- rep(1e-6, n_dim)
+
+  for (i in seq(iterations)) {
```
Should there be a warning if we reach the maximum number of iterations but `error_` is still much larger than the `tolerance`?

pratikunterwegs replied:
What should we consider much larger?

pratikunterwegs replied:
I added an arbitary value of 100x tolerance (typically 1e-4 as default tolerance is 1e-6), as 10x was leading to a number of warnings when iterations < 1e4.

pratikunterwegs replied:
Also added checks to expect warnings for both solvers

Bisaloo commented on R/finalsize_risk_grps.R:
```diff
@@ -10,12 +32,10 @@
-#' work well.
-#' @param adapt_step Boolean, whether the solver step rate should be changed
-#' based on the solver error. Defaults to TRUE.
+#' @param solver Which solver to use. Options are "iterative" or "newton", for
```
```suggestion
#' @param solver Which solver to use. Options are "iterative" (default) or "newton", for
```
